tiri/src/main.rs

```diff
@@ -0,0 +1,121 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Tiri -- TIR interpreter.
+//!
+//! No effort has been made to make this fast.
+
+#[macro_use]
+extern crate log;
+
+use elf;
+use fallible_iterator::FallibleIterator;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::io::Cursor;
+use std::path::{Path, PathBuf};
+use ykpack::{BasicBlockIndex, Body, Decoder, DefId, Pack, Statement, StatementIndex, Terminator};
+
+/// The interpreter maintains a stack of these to keep track of calls.
+struct Frame<'p> {
+    /// A reference to the TIR body corresponding with this `Frame`.
+    body: &'p Body,
+    /// The current execution position in the above TIR body.
+    pc: (BasicBlockIndex, StatementIndex),
+}
+
+impl<'p> Frame<'p> {
+    /// Create a new frame and position its program counter at the beginning of the first block.
+    fn new(body: &'p Body) -> Self {
+        Self { body, pc: (0, 0) }
+    }
+}
+
+/// The interpreter itself.
+/// The struct itself holds only immutable program information.
+struct Interp {
+    /// Maps a `DefId` to the corresponding TIR body.
+    tir_map: HashMap<DefId, Body>,
+    /// The `DefId` of the `main()` function, which serves as the entry point.
+    main_defid: DefId,
+}
+
+impl Interp {
+    /// Create a new interpreter, using the TIR found in the `.yk_tir` section of the binary `bin`.
+    fn new(bin: &Path) -> Self {
+        let ef = elf::File::open_path(&PathBuf::from(bin)).unwrap();
+        let sec = ef.get_section(".yk_tir").unwrap();
+        let mut curs = Cursor::new(&sec.data);
+        let mut dec = Decoder::from(&mut curs);
+
+        let mut tir_map = HashMap::new();
+        let mut main_defid = None;
+        while let Some(pack) = dec.next().unwrap() {
+            let Pack::Body(body) = pack;
+
+            if body.def_path_str == "main" {
+                main_defid = Some(body.def_id.clone());
+            }
+            tir_map.insert(body.def_id.clone(), body);
+        }
+
+        Self {
+            tir_map: tir_map,
+            main_defid: main_defid.expect("Couldn't find main()"),
+        }
+    }
+
+    /// Start interpreting TIR from the `main()` function.
+    fn run(&self) {
+        // Set up the stack to start executing at the entry point.
+        let mut stack = Vec::new();
+        stack.push(Frame::new(&self.tir_map[&self.main_defid]));
+
+        // The main interpreter loop.
+        loop {
+            let cur_frame = stack.last().unwrap();
+            let cur_block = &cur_frame.body.blocks[usize::try_from(cur_frame.pc.0).unwrap()];
+            let block_len = cur_block.stmts.len();
+            let pc_stmt_usize = usize::try_from(cur_frame.pc.1).unwrap();
+
+            if pc_stmt_usize < block_len {
+                let stmt = &cur_block.stmts[pc_stmt_usize];
+                debug!(
```

Comment:
I'm really, really reluctant to merge any code which spews to stdout, even if it's only in debug mode. In my experience, this code sticks around for decades, even when it no longer serves any purpose.

Reply:
I have to disagree with you on this one. There are several reasons I find `debug!` useful, but to cut to the chase and address your main objection:

> this code sticks around for decades, even when it no longer serves any purpose.

These calls will always serve a purpose: they will print a trace of TIR statements leading up to a crash I'm debugging. That's the kind of info that you can't get from a back-trace. When would this not be useful? Especially given that (as you have already noted) they only get compiled in under debug mode, but even then you only see them when a environment variable is set.

I agree that excessive `debug!`s is silly, but strategically placed debug statements save me time overall and there is a precedent to use them in Rust code. The ones in the rust compiler have been especially useful: in the compiler inserting a debug print triggers a lengthy compilation cycle.

Unless you feel very strongly, I'd like to ask politely if we I leave these in :)

Reply:
I do feel *very* strongly about this (sorry). This is how chatty programs start life, and once the precedent is set, it's difficult to stop it happening again and again. If you need them for debugging, that's fine, but that's probably something for your local benefit that isn't useful for anyone else. In my experience, if I think I need to leave debugging statements in, it invariably means that I have an incomplete PR.

> The ones in the rust compiler have been especially useful: in the compiler inserting a debug print triggers a lengthy compilation cycle.

I assume they have some benefit other than making compilation slower? ;)

Reply:
Seems to be some confusion, but I think I know what's going on.

> This is how chatty programs start life, and once the precedent is set, it's difficult to stop it happening again and again. 

I think you have mixed-up `debug!` with `dbg!`? The former *doesn't* lead to chatty programs because you have to opt-in to seeing the messages by setting an environment variable (`RUST_LOG`). You can think of it a bit like `PYPY_LOG`.

You certainly wouldn't want `dbg!` all over the place!

> I assume they have some benefit other than making compilation slower?

No, `debug!` and friends aim to *save* compilation time. Any change to the source tree will trigger recompilation, so if coarse debugging statements are ever-present, often all you need to do to diagnose an issue is set the `RUST_LOG` appropriately.

For example if you wanted info-level information about what the `rustc_driver` crate is doing, you can set `RUST_LOG=rustc_driver=info`. Note that you can filter by crate.

If you still aren't convinced, my last stand (military term that) would be to point out that these macros are really standard. Grepping the compiler:

```
$ grep -R 'debug!' src | wc -l
    2406
```

Reply:
Custer/Barrett's last stand? ;) Here's the problem I have with this: of those 2406, how many are still relevant? My guess is very few. So turning on logging almost certainly overwhelms us with stuff that's irrelevant. That's the problem with putting this stuff in the code base: it never gets GCd.

Reply:
Well the same is true of documentation and comments in my opinion. Just as you have to make sure your comments and docs are up to date, you have to make sure you `debug!`s are too.

I also suspect that as code churns you are forced to review your `debug!`s, since they will surely be the cause of compilation errors during development.

I can see you can't be swayed, so let's move on. Perhaps this will be something we will never agree upon.

Reply:
Yeah, sorry that this is one where it's impossible to find a compromise :/ Thanks for being a good sport about it!